File: queues/queues-custom.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.first = null;
    this.last = null;
    this.size = 0;
  }
  enqueue(val) {
    var newNode = new Node(val);
    if (this.first == null) {
      this.first = newNode;
      this.last = newNode;
    } else {
      this.last.next = newNode;
      this.last = newNode;
    }
    this.size++;
    return this.length;
  }

  dequeue() {
    var temp = this.first;
    if (this.first == null) return null;
    if (this.first === this.last) {
      this.first = null;
      this.last = null;
    } else {
      this.first = this.first.next;
    }
    this.size--;
    return temp.value;
  }
}

var q = new Queue();
q.enqueue(0);
q.enqueue(1);
q.enqueue(2);
q.enqueue(3);

console.log(q);
console.log(q.dequeue())
console.log(q)

File: queues/queues-custom.go
package main

import (
	"fmt"
)

type Node struct {
	value int
	next  *Node
}

type Queue struct {
	first  *Node
	last   *Node
	length int
}

func (q *Queue) traverseQueue() {
	var currentNode = q.first
	for {
		fmt.Println(currentNode)
		if currentNode != q.last {
			currentNode = currentNode.next
		}
	}
}

func (q *Queue) enQueue(n int) int {
	newNode := &Node{value: n}
	if q.first == nil {
		q.first = newNode
		q.last = newNode
	} else {
		q.last.next = newNode
		q.last = newNode
	}
	q.length++
	return q.length
}

func (q *Queue) deQueue() int {
	temp := q.first
	if q.first == nil {
		fmt.Println("Queue is empty")
	}
	if q.first == q.last {
		q.first = nil
		q.last = nil
	} else {
		q.first = q.first.next
	}
	q.length--
	return temp.value
}

func main() {
	q := &Queue{}
	q.enQueue(0)
	q.enQueue(1)
	q.enQueue(2)
	q.enQueue(3)
	q.traverseQueue()
	fmt.Println(q.deQueue())
	fmt.Println(q.deQueue())
}
